### PR TITLE
Add ApplicationRecord for customisations of models

### PR DESCRIPTION
**Pull request by mkllnk**
#### What? Why?

Rails 5 introduced this new class to confine application-specific monkey patches to our models only, and not leak into other libraries using ActiveRecord::Base.

For context: I was adding a new model for #6328 and noticed that Rails generated the `ApplicationRecord` class. I looked up the story behind it and found it quite useful.

https://bigbinary.com/blog/application-record-in-rails-5

Bonus: :fire: +78 −162 lines of code

#### What should we test?

- Sign up and create an enterprise.
- Open a shop (create an order cycle).
- Make sure that you can visit the shop and that the URL matches the enterprise name.

#### Release notes

Changelog Category: Technical changes

We now use the Rails convention to inherit all models from ApplicationRecord. App-specific patches now apply only to ApplicationRecord instead of ActiveRecord::Base. This should lead to fewer unintended side effects.

### CODE REVIEW

**Review by Matt-Yorkley — APPROVED** [review]
Awesome :100:

File: lib/spree/core/permalinks.rb [hunk 1]
```diff
@@ -70,6 +62,3 @@ def save_permalink(permalink_value = to_param)
     end
   end
 end
-
-ActiveRecord::Base.include(Spree::Core::Permalinks)
-ActiveRecord::Relation.include(Spree::Core::Permalinks)
```
Comment (Matt-Yorkley) [inline lib/spree/core/permalinks.rb:75]: I wonder if not including it in `ActiveRecord::Relation` will have any consequences..?

**Review by luisramos0 — APPROVED** [review]
nice :+1:

File: lib/spree/core/permalinks.rb (hunk 1 above)
Comment (mkllnk) [reply lib/spree/core/permalinks.rb:75]: I wondered that, too. But I think that the reason for this were the find_by_param methods. Spree could do something like `Product.in_stock.find_by_param`. But we don't use those any more and I removed them. I couldn't find any usage of the other methods on scopes.

**Comment by filipefurtad0** [comment]
Hey @mkllnk ,

Thank you for the notes on what to test.

I followed your suggestions, and verified:

- signed up and created an enterprise - OK
- opened a shop (create an order cycle) - OK
- make sure that you can visit the shop and that the URL matches the enterprise name:
The shop URL as seen by the customer matches the URL displayed on the admin section, under Dashboard -> Primary details -> Link to shop front - OK

I noticed this PR touches many files! So I thought I just run some quick additional tests, like:
- creating products with stock
- adding them to the order cycle
- creating shipment and payment methods
- checking out as a customer
- checking order states
- capturing (cash) payments

Looks good and ready to go :+1:

**Pull request merged**